### ISSUE TITLE
[🐛 Bug]: Running a single scenario using the scenario line number is not working in @wdio/cucumber-framework version 7.11.0 and later

### ISSUE THREAD

**Issue by nirmalcrafty** [webdriverio/webdriverio#7735]
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md#reporting-new-issues).

### WebdriverIO Version

latest

### Node.js Version

v14.18.1

### Mode

WDIO Testrunner

### Which capabilities are you using?

```typescript
{
"browserName": "chrome",
}
```

### What happened?

Update "@wdio/cucumber-framework" to any version higher than 7.10.1 and notice if you try to ,Run scenario by line number
eg 

wdio.js run wdio.conf.js --spec           
  ./tests/e2e/a.feature:5

It will run the entire feature file with all the scenarios

### What is your expected behavior?

The scenario which is present in the specified line number should be executed

### How to reproduce the bug.

In package.json upgrade the "@wdio/cucumber-framework": "7.16.6", do an npm install and run any scenario in a feature file by passing the below command

say

wdio.js run wdio.conf.js --spec ./tests/e2e/a.feature:5

Note all the scenarios in the a.feature will be running, instead of the scenario present in line number 5

### Relevant log output

```typescript
The results of all the scenarios executed is displayed
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Is there an existing issue for this?

- [X] I have searched the existing issues

**Comment by christian-bromann** [comment]
Thanks for reporting!

This might be a regression due to latest Cucumber update.

Any contributions that resolves the bug are highly appreciated. Please take a look into our [contribution guidelines](https://webdriver.io/docs/contribute) and let us know if you have any questions. Cheers!

**Comment by nirmalcrafty** [comment]
Hi @christian-bromann , only relative paths, are not working, if we send the spec with the absolute path with a colon and line number it works

For eg in Windows

wdio.js run wdio.conf.js --spec c:/tests/e2e/a.feature:5 - This works

but not

wdio.js run wdio.conf.js --spec tests/e2e/a.feature:5

This is because the method getFeatureUriToLinesMapping in pickle_filter.js in @cucumber/cucumber package has now the uris constructed without an absolute path

const uri = path_1.default.normalize(match[1]);

and it used to be 
const uri = path_1.default.resolve(cwd, match[1]);

So maybe we have to send the absolute path constructed  in featurePathsToRun in CucumberAdapter

this._pickleFilter = new Cucumber.PickleFilter({
            cwd: this._cwd,
            featurePaths: featurePathsToRun,
            names: this._cucumberOpts.names,
            tagExpression: this._cucumberOpts.tagExpression
        });

**Comment by christian-bromann** [comment]
Good catch! Mind proposing a PR for it?

**Comment by nirmalcrafty** [comment]
I have submitted a PR, and this is my PR to an open source project https://github.com/webdriverio/webdriverio/pull/7752, I might have some missing permissions I guess

**Comment by Kevin-Irias-vectorsolutions** [comment]
Hello, is there any news about this issue?

**Comment by christian-bromann** [comment]
Doesn't seem like it ;-) any contributions are welcome!

**Comment by aleksyor** [comment]
Hello! I just migrated to wdio 7.25.0 and hit the issue described above. Dos it still persist? And if so - is there any workaround for it?
Thanks a lot in advance!

**Comment by Evotionce** [comment]
@aleksyor I use tags now to mark the tests I want to execute: `--cucumberOpts.tagExpression='@test'`
```
// although it would work with all feature files selected in spec, it is much quicker to just use the specific feature file with your tag
./wdio.conf.js --cucumberOpts.tagExpression='@test' --spec <your spec>

// in your feature file
@test 
Scenario: This is my scenario I want to test
```

Btw any tag name would work. Hope it helps!

**Comment by aleksyor** [comment]
Thanks a lot, @Evotionce. This is pretty good workaround.

**Comment by christian-bromann** [comment]
This is a pretty well scoped issue that can be an ideal first issue for someone who hasn't contributed to the project yet. We contributor offer [1:1 sessions](https://webdriver.io/community/openofficehours) to help if folks have issues or questions working on certain tickets. I would be nice to see a bug fix for this rather than relying on "workarounds".

**Comment by mikesalvia** [comment]
@christian-bromann in my independent verification I found this issue popping up between 7.20.3 (last working) and 7.20.6 (first break) of @wdio/cucumber-framework.

**Comment by mikesalvia** [comment]
Fairly confident this is caused by #8432.

**Comment by christian-bromann** [comment]
Thanks for the investigation, any fix would be much appreciated.

**Comment by hammzj** [comment]
There's been a lot of back and forth on this issue and it seems like there's something missing around the expectations around the order of operations for filtering tests. With the PR #8432, I had [this comment](https://github.com/webdriverio/webdriverio/issues/8376#issuecomment-1146265155) that tried to define what the expected order _should_ be, but now it seems like that might not be the case for some other options. It might be a time to map out what the hierarchy is for running Cucumber scenarios.

**Comment by mikesalvia** [comment]
@hammzj Totally understand! What is weird is it may not be that commit that is the errant code but that is exposed a flaw in logic in @wdio/cucumber-framework, like relying on `specs` and `spec`.

What order of operations for Cucumber scenario filtering is there? Keeping in mind we have 3 possible inputs for `--spec`:

* Cucumber Feature path
* Cucumber Feature path with Scenario line number
* Mocha/Jasmine test files

@christian-bromann Working on figuring that out now. I am hoping that the V7 fix would carry over to V8 with minimal issues.

**Comment by mikesalvia** [comment]
@hammzj sorry about that I saw your proposed order. Going to look at how line numbers fit into that.

**Comment by hammzj** [comment]
@mikesalvia, I posted an outline for a solution here:
https://github.com/webdriverio/webdriverio/pull/8432#issuecomment-1419216928
Would love some feedback!

**Comment by christian-bromann** [comment]
@mikesalvia any progress on this? Happy to help if any questions arise.

**Comment by mikesalvia** [comment]
@christian-bromann what is the use case for specs in caps which is what @hammzj was fixing here I think.

I am trying to configure all paths but I am stuck on that.

**Comment by Grubsniff** [comment]
Hi just wondering if there's been any update to this issue or other fixes that may address it at all?

We recently started trying to upgrade to v8 of wdio packages, from v7 and have found this issue ^ with executing cucumber scenario via line number.

**Comment by christian-bromann** [comment]
> Hi just wondering if there's been any update to this issue or other fixes that may address it at all?

It doesn't seem to me anyone is actively working on this so feel free to pick this up. Let us know if you have any questions.

**Comment by mikesalvia** [comment]
I plan to fix this both for v7 and v8, v8 first I just need to carve out some time to work on it.

**Comment by christian-bromann** [comment]
Thanks @mikesalvia , let me know if I can help along the way.

**Comment by hammzj** [comment]
Hey y'all, I've gotten some time to look at some WDIO issues and am trying to diagnose what is going on here. I think this is an issue with the @wdio/cucumber-framework package, instead.

There is a `pickleFilter` that is supposed [to filter out scenarios based on tags, feature names, and line numbers](https://github.com/cucumber/cucumber-js/blob/2fd424aa7127a6449291015ad326f555ed957203/src/pickle_filter.ts#L47). I can see that the [pickleFilter in the cucumber-framework](https://github.com/hammzj/webdriverio/blob/b03188f55479f936d9ca899967883d929cde191c/packages/wdio-cucumber-framework/src/index.ts#L60C14-L60C27) _does_ pass in a `lineFilter` for a given Cucumber feature with line numbers (like `./tests/e2e/a.feature:5`). However, when it performs the filtering, it's not accounting for the `lineFilter` and is returning all pickleIds for each scenario in the file.

The problem probably exists somewhere in the Cucumber event listener's method, `getPickleIds` ([See here](https://github.com/hammzj/webdriverio/blob/b03188f55479f936d9ca899967883d929cde191c/packages/wdio-cucumber-framework/src/cucumberEventListener.ts#L376)).

I could be completely inaccurate here, but I think it's a good start to look at how the WDIO Cucumber framework utilizes the pickle filtering here.

**Comment by hammzj** [comment]
Ha, I think I found it.

The `pickleFilter.lineFilter` has a method named [`matchesAnyLine`](https://github.com/cucumber/cucumber-js/blob/2fd424aa7127a6449291015ad326f555ed957203/src/pickle_filter.ts#L98):

```ts
  matchesAnyLine({ gherkinDocument, pickle }: IMatchesAnyLineRequest): boolean {
    const uri = path.normalize(pickle.uri)
    const linesToMatch = this.featureUriToLinesMapping[uri]
    if (doesHaveValue(linesToMatch)) {
      const gherkinScenarioLocationMap =
        getGherkinScenarioLocationMap(gherkinDocument)
      const pickleLines = new Set(
        pickle.astNodeIds.map(
          (sourceId) => gherkinScenarioLocationMap[sourceId].line
        )
      )
      const linesIntersection = linesToMatch.filter((x) => pickleLines.has(x))
      return linesIntersection.length > 0
    }
    return true
  }
}
```

In this method, it tries to get the line numbers from `this.featureUriToLinesMapping` using the `pickle.uri`.
What is happening is that WDIO is passing in the **full, absolute path** as the `pickle.uri`, but that `this.featureUriToLinesMapping` is set to the `--spec` option's path, which is most likely a relative path.

For example, If I run `yarn wdio --spec ./tests/e2e/a.feature:5`, then this is what happens:
* The ConfigParser will set the pickleFilter to be:`this.featureUriToLinesMapping = {"tests/e2e/a.feature": [5]}`
* The WDIO Cucumber framework sets the `pickle.uri` as the full path, like "/Users/hammz/Documents/GitHub/my_project/tests/e2e/a.feature",
* No lines will ever be matched because `this.featureUriToLinesMapping[pickle.uri]` will be the same as `this.featureUriToLinesMapping["./GitHub/my_project/tests/e2e/a.feature"]`, which returns `undefined`.

What I think needs to happen is that the [initial pickleFilter in the CucumberAdapter](https://github.com/hammzj/webdriverio/blob/a0865fb9c7a07662530e1d6190efd2a103148e00/packages/wdio-cucumber-framework/src/index.ts#L86) needs to pass in the absolute paths of each feature to the pickleFilter.

I can try to take this up, but wanted to provide a potential fix if anyone wants to try it out!

**Comment by christian-bromann** [comment]
Awesome 🎉 would you like to raise a PR for this?

**Comment by hammzj** [comment]
Sure, might be a week or so but I'll try.

**Comment by esaari** [comment]
I'll update the dependency in the rerun service as soon as it's released!

**Comment by hammzj** [comment]
You're lucky. I put some work in today for y'all 😉  https://github.com/webdriverio/webdriverio/pull/10772

**Comment by hammzj** [comment]
I'd like some more validation from other people before we merge this -- I don't want to introduce another breaking change if I did so already!

**Comment by hammzj** [comment]
Hey @esaari , my changes are available in [v8.13.9](https://github.com/webdriverio/webdriverio/releases/tag/v8.13.9). Can you check and see if this can be closed out if it is working for you?

**Comment by christian-bromann** [comment]
I will be closing this ticket. If anyone wants to raise a PR to apply changes made in #10772 to the `v7` branch I am happy to merge and release that for everyone still using v7.
